### DISCUSSION TITLE
How to configure zarr to store/compress arrays deterministically?

### DISCUSSION THREAD

**Discussion by jjyyxx** [zarr-developers/zarr-python#1519] [Q&A]
I recently switched from simple `np.savez` to zarr for more flexibility with generating a big dataset. The only thing currently missing is that the compressed array is not deterministic given same input. This leads to certain difficulty in our version maintenance. So, where does the randomness come from? Is it possible to configure zarr to store/compress arrays deterministically, like setting a seed?

**Comment by rabernat** [comment]
> The only thing currently missing is that the compressed array is not deterministic given same input.

Can you explain what you mean by "not deterministic"? Most of the popular zarr compressors are lossless compressors. Can you share the details of your configuration (what dtype, compressor and parameters) and perhaps a reproducible example.

**Reply by jjyyxx** [reply]
Sorry for my inclarify. zarr has already been very handy at solving my problems. It is of course lossless in the sense of save-then-load round-trip. 

But, take `DirectoryStore` for example. By "not deterministic", I mean the arrays written to the disk (after compression) is not byte-by-byte identical across multiple runs. In other words, `diff -r .../1.zarr/ .../2.zarr/` shows that every array is different (but `.zarray` and `.zgroup` files are identical).

**Reply by rabernat** [reply]
I'm not able to reproduce this behavior with a simple example.

```python
import zarr

store1 = zarr.storage.MemoryStore()
store2 = zarr.storage.MemoryStore()

arr1 = zarr.create(store=store1, shape=10, chunks=10)
arr1[:] = 2

arr2 = zarr.create(store=store2, shape=10, chunks=10)
arr2[:] = 2

# stores are identical
assert(dict(store1) == dict(store2))
```

Can you please share a [minimal complete verifiable example](https://matthewrocklin.com/minimal-bug-reports) that a developer can copy and paste in order to verify your issue?

**Reply by jjyyxx** [reply]
Unless it is platform- or hardware-dependent, larger arrays seem to cause non-determinism.

```python
import zarr

def test_determinism_once(size: int):
    store1 = zarr.MemoryStore()
    store2 = zarr.MemoryStore()

    arr1 = zarr.create(store=store1, shape=size, chunks=False)
    arr1[:] = 2

    arr2 = zarr.create(store=store2, shape=size, chunks=False)
    arr2[:] = 2

    return dict(store1) == dict(store2)

def test_determinism(size: int, n: int = 1000):
    percentage = sum(test_determinism_once(size) for _ in range(n)) / n
    print(f'{percentage:>6.1%} of {n} runs are deterministic for size {size}')

test_determinism(1000)
test_determinism(262143)
test_determinism(262144)  # 2**18
test_determinism(1000000)
```
prints
```
100.0% of 1000 runs are deterministic for size 1000
100.0% of 1000 runs are deterministic for size 262143
 53.6% of 1000 runs are deterministic for size 262144
  0.0% of 1000 runs are deterministic for size 1000000
```
Guess such non-determinism may be multi-threading related.

**Reply by rabernat** [reply]
Thanks for this great example!

It appears to be compressor dependent. The default compressor is `Blosc(cname='lz4', clevel=5, shuffle=SHUFFLE, blocksize=0)`

If I try with a different compressor, I get 100% determinism.

```python
from numcodecs import Zstd
compressor = Zstd(level=1)
arr1 = zarr.create(store=store1, shape=size, chunks=False, compressor=compressor)
arr2 = zarr.create(store=store2, shape=size, chunks=False, compressor=compressor)
```

(You'll also note that zstd is much slower!)

It could be that the lz4 compression algorithm itself is non-deterministic or involves some sort of seed. I would do some research on that. It's not specifically a zarr issue I don't think. Zarr is just sending the bytes to [numcodecs](https://numcodecs.readthedocs.io/en/stable/).

edit: also possibly related to blosc multithreading.

**Reply by jjyyxx** [reply]
A bit of context: I was originally using `np.savez_compressed` to store my dataset, but it uses `ZIP_DEFLATED` which is extremely slow for large arrays and does not support chunks. So, I switched to `zarr` to get over 10 times speed up. The size on disk is 20% larger compared to `ZIP_DEFLATED`, but I'm OK with that. (So, switching back to `zstd` is not an option to me)

**Reply by rabernat** [reply]
Try configuring blosc to only use a single thread.

**Reply by ktpolanski** [reply]
I just ran into something similar and found this discussion very helpful. I initially spotted the nondeterminism in AnnData's zarr storage of a small data frame, and similarly noticed it escalating as I tried a larger input later. I can confirm that running blosc with a single thread led to everything behaving deterministically.

I set up the single threaded blosc like so, just in case there's a better way to do this:
```python3
from numcodecs import Blosc
import numcodecs.blosc
numcodecs.blosc.set_nthreads(1)
compressor = Blosc()
```
